### ISSUE TITLE
Insert into... select

### ISSUE THREAD

**Issue by pulli23** [knex/knex#4700]
# Environment

Knex version: 0.21 (but also 0.95)
Database + version: postgresql
OS: linux

# Feature discussion / request

1. Explain what is your use case

Well in sql there is the statement 
```
INSERT INTO table2
SELECT * FROM table1
WHERE condition;
```
In knex while one could use subqueries it then becomes near impossible to select the columns that are provided (and hence the database expects even automatic generated rows like ID to be provided).

2. Explain what kind of feature would support this
A simple example would be:

```
      const query = knex.into('my_table').insert((q) => {
        q
          .from('other_table')
          .select([
            'other_table.field as translated_name',
          ])
      })
```
This would then select the correct columns (similarly to how knex translates the array of data to an object) for the insert clause. The topic has been discussed way back in the past here: https://github.com/knex/knex/issues/1323 -- however it seems to be close without a solution.

3. Give some API proposal, how the feature should work

```
      const query = knex.into('my_table').insert((q) => {
        q
          .from('other_table')
          .select([
            'other_table.field as translated_name',
          ])
      })
```

Would give a query like:

```
INSERT INTO "my_table" ("translated_name")
SELECT "other_table.field" as translated_name
FROM "other_table"
```

**Comment by elhigu** [comment]
It can be done like this:

```
knex('my_table').insert(knex('other_table').select(['other_table.field as translated_name']))
```

https://runkit.com/embed/t5eer0p1x55u

**Comment by pulli23** [comment]
@elhigu no as I said that would give an sql like:

```
insert into "my_table" select "other_table"."field" as "translated_name"
```
So it tries to insert into the first column of the `my_table` (which could be the ID column) the selected field.

**Comment by elhigu** [comment]
Ah right.  I’ll add later on laptop how to do it currently. If special syntax would be added, then syntax should ve probably something else than the one proposed here, since if passing query builder / builder callback default behavior is changed that might be confusing.

**Comment by elhigu** [comment]
So this is how it would be written currently:

``` 
knex('my_table').insert(knex.raw("(??) ??", [
  "translated_name",
  knex('other_table').select(['other_table.field as translated_name'])
]))
```

https://runkit.com/embed/u3xjbebskd4s

**Comment by pulli23** [comment]
While that would be the case, it uses `knex.raw` and really makes the whole query hard to read. (Especially if you have to layer `knex.raw` statements to add static data as well.

A use case for example would be a request where one wishes to add "managers" to an "organizations", the organization ID (which is public) is given, but to add a manager the user has searched/given the username of the new manager.

Then in the `managerorganization` joint table those would be added.

**Comment by elhigu** [comment]
Maybe this could be a place for implementing `.insertFromSelect()` knex extension? 

How to extend knex is described near here in the docs https://knexjs.org/#Builder-queryContext (documentation didn't have direct link to that subheader)

Method sounds quite opinionated (but also useful). If it reaches huge userbase something like that probably could be added to knex core also, but for now I believe it is better to implement as a separate npm package and knex can add link and description to it in the ecosystem docs.

Actually I would prefer most of the new features to be first implemented and ironed out outside of knex before bringing anything to main repo. That way it is a lot easier to make changes to it if the first implementation / API was not as good as it was thought to be.

**Comment by jpike88** [comment]
INSERT ... SELECT is part of the SQL standard and widely supported by major SQL databases.

Knex should support it properly, it carries a serious risk of sql injection attacks

**Comment by mercmobily** [comment]
@jpike88 I agree. I am adding this as a feature request. We have a lot of those, but we are hoping to go through them soon enough.
